src/backend/commands/copyfrom.c

```diff
@@ -666,6 +682,777 @@ CopyMultiInsertInfoStore(CopyMultiInsertInfo *miinfo, ResultRelInfo *rri,
+				 Datum *values,
+				 bool *nulls)
+{
+	TupleDesc	tupDesc;
+	AttrNumber	num_phys_attrs;
+	ListCell   *cur;
+	char	   *field[6];
+	FmgrInfo   *in_functions = cstate->in_functions;
+	Oid		   *typioparams = cstate->typioparams;
+	List	   *attnumlist = cstate->qd_attnumlist;
+	pg_time_t	stampTime = (pg_time_t) time(NULL);
+	char		lastModified[128];
+	char	*encode_file;
+	int		encode_file_len;
+
+	encode_file_len = pg_b64_enc_len(buf->len);
+	encode_file = (char *) palloc0(encode_file_len + 1);
+
+	encode_file_len = pg_b64_encode(buf->data, buf->len, encode_file, encode_file_len);
+	if (encode_file_len < 0)
+	{
+		elog(ERROR, "base 64 encode failed in copy from directory table");
+	}
+	encode_file[encode_file_len] = '\0';
+	if (buf->data)
+	{
+		pfree(buf->data);
+	}
+
+	pg_strftime(lastModified, sizeof(lastModified),
+				"%Y-%m-%d %H:%M:%S",
+				pg_localtime(&stampTime, log_timezone));
+
+	tupDesc = RelationGetDescr(cstate->rel);
+	num_phys_attrs = tupDesc->natts;
+
+	MemSet(values, 0, num_phys_attrs * sizeof(Datum));
+	MemSet(nulls, false, num_phys_attrs * sizeof(bool));
+
+	field[0] = relativePath; /* relative_path */
+	field[1] = psprintf(INT64_FORMAT, fileSize); /* size */
+	field[2] = lastModified; /* last_modified */
+	field[3] = md5sum; /* md5sum */
+	field[4] = tags; /* tags */
+	if (tags == NULL)
+		nulls[4] = true;
+	field[5] = encode_file;
+
+	/* Loop to read the user attributes on the line. */
+	foreach(cur, attnumlist)
+	{
+		int    attnum = lfirst_int(cur);
+		int    m = attnum - 1;
+		char *value;
+		Form_pg_attribute att = TupleDescAttr(tupDesc, m);
+
+		value = field[m];
+
+		values[m] = InputFunctionCall(&in_functions[m],
+									  value,
+									  typioparams[m],
+									  att->atttypmod);
+	}
+}
+
+/*
+ * Copy From file to directory table.
+ */
+static uint64
+CopyFromDirectoryTable(CopyFromState cstate)
+{
+	ResultRelInfo *resultRelInfo;
+	ResultRelInfo *target_resultRelInfo;
+	EState	   *estate = CreateExecutorState();
+	TupleTableSlot *myslot = NULL;
+	StringInfoData	buf;
+	ExprContext *econtext;
+	int			bytesRead;
+	char		hexMd5Sum[256];
+	char		buffer[DIR_FILE_BUFF_SIZE];
+	int64		processed = 0;
+	int64		fileSize = 0;
+	CdbCopy	   *cdbCopy = NULL;
+	char	   *orgiFileName;
+	char	   *relaFileName;
+	TupleDesc	tupdesc;
+	unsigned int targetSeg;
+	DirectoryTable     *dirTable;
+	pg_cryptohash_ctx  *hashCtx;
+	uint8 md5Sum[MD5_DIGEST_LENGTH];
+	GpDistributionData *distData = NULL; /* distribution data used to compute target seg */
+
+	/*
+	 * We need a ResultRelInfo so we can use the regular executor's
+	 * index-entry-making machinery.  (There used to be a huge amount of code
+	 * here that basically duplicated execUtils.c ...)
+	 */
+	ExecInitRangeTable(estate, cstate->range_table);
+	resultRelInfo = target_resultRelInfo = makeNode(ResultRelInfo);
+	ExecInitResultRelation(estate, resultRelInfo, 1);
+
+	ExecOpenIndices(resultRelInfo, false);
+
+	/* Prepare to catch AFTER triggers. */
+	AfterTriggerBeginQuery();
```

Comment:
seem to support triggers ?

Reply:
No, should not support, I will disallow those behaviors.

Reply:
In my opinion, maybe supporting triggers is better.

Reply:
ok, add a test cases.

Reply:
OK